### PR TITLE
Optimization for function sets 

### PR DESCRIPTION
**Pull request by MikeStall**
Optimize the pattern where:
- a host calls eval many times in a loop. 
- but using the same function set in all cases, as determined by just PowerFxConfig (with some enables).

This PR is an optimization and there shouldn't be any functional changes. However - we should review some of the behavior around TexlFunctionSet to enable us to pursue more comprehensive optimizations. 

Today, every call to Check() will compute the function set - which involves merging the function set from builtin functions plus the extra ones that were enabled.  This is expensive because TexlFunctionSet will "flatten" and copy all the functions into a single list. (Contrast to how we compose SymbolTables - which is entirely deferred). 

The optimization is to allow composing that once and caching it. But also to be sure we have correct invalidation strategy if anything changes. Added lots more tests as I discovered interesting corner cases. 

This makes the benchmark use 1/10 the working set and run 3x faster.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 1]
```diff
@@ -52,8 +52,13 @@ internal class TexlFunctionSet
 
         internal IEnumerable<string> Enums => _enums;
 
-        // Return an empty TexlFucntion set that can never be added to. 
-        internal static TexlFunctionSet Empty()
+        // Return an singleton for an empty TexlFunction set that can never be added to. 
+        internal static readonly TexlFunctionSet _empty = MakeEmpty();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:56]: this saved about 1kb per loop.

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 2]
```diff
@@ -98,6 +103,35 @@ internal TexlFunctionSet(IEnumerable<TexlFunction> functions)
             }
         }
 
+        // This will compose multiple function sets into a single one. 
+        // The assumption is that all incoming function sets are unchanged for the lifetime of the returned set.
+        // Hence it doesn't matter where this shares references or copies the values. 
+        internal static TexlFunctionSet Compose(IEnumerable<TexlFunctionSet> functionSets)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:109]: This is critical. 
Each symbol table has a function set, but most are empty. 
It's also key that the caller of this method knows the symbols won't change on us - this is true because this is only called during a binding pass (which operates on a fixed set of symbols - and has safety checks to throw if the symbols are changed during binding).

File: src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs [hunk 3]
```diff
@@ -163,14 +163,21 @@ private protected INameResolver CreateResolverInternal(out ReadOnlySymbolTable s
...
         /// Get a combined engine symbol table, including builtins and config. 
         /// </summary>
         /// <returns></returns>
         public ReadOnlySymbolTable GetCombinedEngineSymbols()
         {
-            var symbols = ReadOnlySymbolTable.Compose(EngineSymbols, SupportedFunctions, Config.SymbolTable, PrimitiveTypes);
+            // Compose all the symbol tables most likely to have functions into a single 
+            // symbol table and then cache that. 
+            // That will cache unifying into a single TexlFunctionSet - which is the most expensive part. 
+            var functionList = _functionListCache.GetComposedCached(SupportedFunctions, Config.SymbolTable);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs:178]: This is critical.  This combines the symbol tables that have non-empty function sets into a single table, which can then be cached.  And being a single table, then the Compose() operator above becomes trivial and fast.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/ConfigTests.cs [hunk 4]
```diff
@@ -788,6 +788,23 @@ public async Task DefaultFunctions()
...
+        {
+            var config = new PowerFxConfig();
+            config.AddFunction(new Func1Function());
+            
+            var engine = new Engine(config);
+           
+            var result = engine.Check("Func1(3)"); // 6
+            Assert.True(result.IsSuccess);
+
+            config.AddFunction(new MultiplyFunction());
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/ConfigTests.cs:803]: This line is key - ensuring that even if the config is mutated, the invalidation is still working.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedSymbolTableCache.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                }
+            }
+        }
+
+        public ReadOnlySymbolTable GetComposedCached(params ReadOnlySymbolTable[] tables)
+        {
+            // This is a conceptually read-only operation, so it could be called on multiple threads. 
+            // CA2002 - Lock(this) is safe since it's a private object that doesn't cross AD boundaries.
+#pragma warning disable CA2002 // Do not lock on objects with weak identity
+            lock (this)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedSymbolTableCache.cs:55]: >lock (this) [](http://example.com/codeflow?start=12&length=11)

why not just lock on private object to avoid this warning and keep it simple?

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs [hunk 6]
```diff
@@ -387,6 +387,76 @@ public void Threading()
...
+        // It's ok to pass nulls to GetComposedCached. 
+        [Fact]
+        public void CacheNullOk()
+        {
+            var table1 = new SymbolTable { DebugName = "Table1" };
+            table1.AddConstant("c1", FormulaValue.New("constant"));
+            ReadOnlySymbolTable nullTable = null;
+            
+            ComposedSymbolTableCache cache = new ComposedSymbolTableCache();
+            var c1 = cache.GetComposedCached(table1, nullTable);
```
Comment (jas-valgotar) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs:453]: >cache.GetComposedCached(table1, nullTable); [](http://example.com/codeflow?start=21&length=43)

Will we null ref if?
cache.GetComposedCached(nullTable);

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs [hunk 7]
```diff
@@ -387,6 +387,76 @@ public void Threading()
...
+            var c1 = cache.GetComposedCached(table1, table2);
+
+            // Same args, cache returns identical instance
+            {
+                var c1b = cache.GetComposedCached(table1, table2);
+
+                Assert.True(object.ReferenceEquals(c1, c1b));
+            }
+
+            // Mutating an existing talbe is ok - ComposedSymbolTable will catch that
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs:406]: >talbe [](http://example.com/codeflow?start=36&length=5)

nit: typo

**Review by jas-valgotar — DISMISSED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedSymbolTableCache.cs (hunk 5 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedSymbolTableCache.cs:55]: will update

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs (hunk 6 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/SymbolTableTests.cs:453]: should be fine. null array will nullref, but that's ok - developer error.   null items in an array are fine.   I'll add this test.

**Review by jas-valgotar — APPROVED** [review]
:shipit:

**Review by anderson-joyle — APPROVED** [review]
:shipit:
